File: gateway/pkg/payload/payload.go
// Package payload constructs and serialises the REP JSON payload
// that is injected into HTML documents.
//
// See REP-RFC-0001 §4.3 (HTML injection) and §8.1 (wire format).
package payload

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/ruach-tech/rep/gateway/internal/config"
	repcrypto "github.com/ruach-tech/rep/gateway/internal/crypto"
)

// Payload is the JSON structure injected into HTML documents.
// See REP-RFC-0001 §8.1 for the schema.
type Payload struct {
	Public    map[string]string `json:"public"`
	Sensitive string            `json:"sensitive,omitempty"`
	Meta      Meta              `json:"_meta"`
}

// Meta contains metadata about the payload.
type Meta struct {
	Version    string `json:"version"`
	InjectedAt string `json:"injected_at"`
	Integrity  string `json:"integrity"`
	KeyEndpoint string `json:"key_endpoint,omitempty"`
	HotReload  string `json:"hot_reload,omitempty"`
	TTL        int    `json:"ttl"`
}

// Builder constructs REP payloads from classified variables.
type Builder struct {
	keys       *repcrypto.Keys
	version    string
	hotReload  bool
}

// NewBuilder creates a payload builder with the given cryptographic keys.
func NewBuilder(keys *repcrypto.Keys, version string, hotReload bool) *Builder {
	return &Builder{
		keys:      keys,
		version:   version,
		hotReload: hotReload,
	}
}

// Build constructs the full REP payload from classified variables.
//
// This performs the following steps per §4.2 (startup sequence steps 7–9):
//  1. Computes HMAC integrity token over public + sensitive data.
//  2. Encrypts sensitive variables using AES-256-GCM.
//  3. Constructs the JSON payload object.
func (b *Builder) Build(vars *config.ClassifiedVars) (*Payload, error) {
	publicMap := vars.PublicMap()
	sensitiveMap := vars.SensitiveMap()

	// Step 1: Compute the integrity token over the public vars only.
	// This value is stored in _meta.integrity AND used as AAD for AES-GCM
	// encryption, so the SDK can use _payload._meta.integrity as AAD for
	// decryption (per §8.2). A single-pass approach avoids the circular
	// dependency that arises when trying to include the encrypted blob in
	// its own encryption AAD.
	integrity := repcrypto.ComputeIntegrity(publicMap, "", b.keys.HMACSecret)

	// Step 2: Encrypt sensitive variables, binding them to integrity via AAD.
	var sensitiveBlob string
	if len(sensitiveMap) > 0 {
		var err error
		sensitiveBlob, err = repcrypto.EncryptSensitive(sensitiveMap, b.keys.EncryptionKey, integrity)
		if err != nil {
			return nil, fmt.Errorf("encrypting sensitive vars: %w", err)
		}
	}

	// Construct the payload.
	p := &Payload{
		Public:    publicMap,
		Sensitive: sensitiveBlob,
		Meta: Meta{
			Version:    b.version,
			InjectedAt: time.Now().UTC().Format(time.RFC3339Nano),
			Integrity:  integrity,
			TTL:        0,
		},
	}

	// Add session key endpoint if sensitive vars exist.
	if len(sensitiveMap) > 0 {
		p.Meta.KeyEndpoint = "/rep/session-key"
	}

	// Add hot reload endpoint if enabled.
	if b.hotReload {
		p.Meta.HotReload = "/rep/changes"
	}

	return p, nil
}

// MarshalJSON serialises the payload to JSON bytes.
func (p *Payload) MarshalJSON() ([]byte, error) {
	// Use an alias to avoid infinite recursion.
	type PayloadAlias Payload
	return json.Marshal((*PayloadAlias)(p))
}

// ToJSON serialises the payload and returns the bytes.
func (p *Payload) ToJSON() ([]byte, error) {
	return json.Marshal(p)
}

// ScriptTag generates the full <script> block for HTML injection.
//
// Per §4.3, the block has:
//   - id="__rep__" for SDK discovery
//   - type="application/json" to prevent execution
//   - data-rep-version for protocol version
//   - data-rep-integrity for SRI verification
func (p *Payload) ScriptTag() (string, error) {
	jsonBytes, err := p.ToJSON()
	if err != nil {
		return "", fmt.Errorf("serialising payload: %w", err)
	}

	sri := repcrypto.ComputeSRI(jsonBytes)

	return fmt.Sprintf(
		`<script id="__rep__" type="application/json" data-rep-version="%s" data-rep-integrity="%s">%s</script>`,
		p.Meta.Version,
		sri,
		string(jsonBytes),
	), nil
}

File: plugins/next/src/payload.ts
import {
  computeIntegrity,
  encryptSensitive,
  computeSRI,
  goEscapeJSON,
  type Keys,
} from './crypto.js';
import type { ClassifiedVars } from './env.js';

export interface PayloadResult {
  json: string;
  scriptTag: string;
  sri: string;
}

export interface PayloadOptions {
  version: string;
  hotReload?: boolean;
}

/**
 * Build the REP payload JSON and script tag.
 * Matches gateway/pkg/payload/payload.go:Build + ScriptTag exactly.
 *
 * JSON field order must match Go struct serialization:
 *   { "public": {...}, "sensitive": "...", "_meta": {...} }
 *
 * All string values get Go HTML escaping (< > & → \u003c \u003e \u0026).
 */
export function buildPayload(
  classified: ClassifiedVars,
  keys: Keys,
  options: PayloadOptions,
): PayloadResult {
  const publicMap = classified.public;
  const sensitiveMap = classified.sensitive;
  const hasSensitive = Object.keys(sensitiveMap).length > 0;

  // Step 1: Compute integrity over public vars only (empty sensitive blob).
  // This matches Go's payload.go — integrity is computed BEFORE encryption,
  // and the integrity token is used as AAD for AES-GCM.
  const integrity = computeIntegrity(publicMap, '', keys.hmacSecret);

  // Step 2: Encrypt sensitive vars if present.
  let sensitiveBlob = '';
  if (hasSensitive) {
    sensitiveBlob = encryptSensitive(sensitiveMap, keys.encryptionKey, integrity);
  }

  // Step 3: Build the payload JSON manually to match Go's field order.
  // Go's json.Marshal produces fields in struct declaration order:
  //   public, sensitive (omitempty), _meta
  const injectedAt = new Date().toISOString();

  // Build _meta object
  const metaParts: string[] = [
    `"version":${goEscapeJSON(JSON.stringify(options.version))}`,
    `"injected_at":${goEscapeJSON(JSON.stringify(injectedAt))}`,
    `"integrity":${goEscapeJSON(JSON.stringify(integrity))}`,
  ];

  if (hasSensitive) {
    metaParts.push(`"key_endpoint":"/rep/session-key"`);
  }

  if (options.hotReload) {
    metaParts.push(`"hot_reload":"/rep/changes"`);
  }

  metaParts.push(`"ttl":0`);

  // Build public object with sorted keys + Go escaping
  const publicSorted = Object.keys(publicMap).sort();
  const publicParts = publicSorted.map(
    k => `${goEscapeJSON(JSON.stringify(k))}:${goEscapeJSON(JSON.stringify(publicMap[k]))}`,
  );
  const publicJSON = `{${publicParts.join(',')}}`;

  // Assemble top-level payload
  const parts: string[] = [`"public":${publicJSON}`];

  if (hasSensitive) {
    parts.push(`"sensitive":${goEscapeJSON(JSON.stringify(sensitiveBlob))}`);
  }

  parts.push(`"_meta":{${metaParts.join(',')}}`);

  const json = `{${parts.join(',')}}`;
  const jsonBytes = Buffer.from(json, 'utf-8');
  const sri = computeSRI(jsonBytes);

  const scriptTag =
    `<script id="__rep__" type="application/json" data-rep-version="${options.version}" data-rep-integrity="${sri}">${json}</script>`;

  return { json, scriptTag, sri };
}
